docs/architecture/adr-070-collections-state-layer.md

```diff
@@ -0,0 +1,117 @@
+# ADR 070: Collections, a simplified storage layer for cosmos-sdk modules.
+
+## Changelog
+
+* 30/11/2022: PROPOSED
+
+## Status
+
+PROPOSED - Implemented
+
+## Abstract
+
+We propose a simplified module storage layer which leverages golang generics to allow module developers to handle module
+storage in a simple and straightforward manner, whilst offering safety, extensibility and standardisation.
+
+## Context
+
+Module developers are forced into manually implementing storage functionalities in their modules, those functionalities include
+but are not limited to:
+
+- Defining key to bytes formats.
+- Defining value to bytes formats.
+- Defining secondary indexes.
+- Defining query methods to expose outside to deal with storage.
+- Defining local methods to deal with storage writing.
+- Dealing with genesis imports and exports.
+- Writing tests for all the above.
+
+
+This brings in a lot of problems:
+- It blocks developers from focusing on the most important part: writing business logic.
+- Key to bytes formats are complex and their definition is error-prone, for example:
+  - how do I format time to bytes in such a way that bytes are sorted?
+  - how do I ensure when I don't have namespace collisions when dealing with secondary indexes?
+- The lack of standardisation makes life hard for clients, and the problem is exacerbated when it comes to providing proofs for objects present in state. Clients are forced to maintain a list of object paths to gather proofs.
+
+### Current Solution: ORM
+
+The current SDK proposed solution to this problem is [ORM](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-055-orm.md).
+Whilst ORM offers a lot of good functionality aimed at solving these specific problems, it has some downsides:
+- It requires migrations.
+- It uses the newest protobuf golang API, whilst the SDK still mainly uses gogoproto. 
+- Integrating ORM into a module would require the developer to deal with two different golang frameworks (golang protobuf + gogoproto) representing the same API objects.
```

Comment:
This is a big downside. It is unfortunate we haven't gotten further with ripping out gogo

Reply:
oh